File: apps/frontend/src/adapters/shared/portfolios.ts
import type { NewPortfolio, PortfolioWithAccounts } from "@/lib/types";
import { invoke, logger } from "./platform";

export const getPortfolios = async (): Promise<PortfolioWithAccounts[]> => {
  try {
    return await invoke<PortfolioWithAccounts[]>("get_portfolios");
  } catch (error) {
    logger.error("Error fetching portfolios.");
    throw error;
  }
};

export const createPortfolio = async (portfolio: NewPortfolio): Promise<PortfolioWithAccounts> => {
  try {
    return await invoke<PortfolioWithAccounts>("create_portfolio", { portfolio });
  } catch (error) {
    logger.error("Error creating portfolio.");
    throw error;
  }
};

export const updatePortfolioEntry = async (
  portfolio: PortfolioWithAccounts,
): Promise<PortfolioWithAccounts> => {
  try {
    return await invoke<PortfolioWithAccounts>("update_portfolio_entry", { portfolio });
  } catch (error) {
    logger.error("Error updating portfolio.");
    throw error;
  }
};

export const deletePortfolio = async (portfolioId: string): Promise<void> => {
  try {
    await invoke<void>("delete_portfolio_entry", { portfolioId });
  } catch (error) {
    logger.error("Error deleting portfolio.");
    throw error;
  }
};

File: apps/tauri/src/commands/portfolios.rs
use std::sync::Arc;

use tauri::State;

use crate::context::ServiceContext;
use wealthfolio_core::portfolios::{NewPortfolio, PortfolioUpdate, PortfolioWithAccounts};

#[tauri::command]
pub async fn get_portfolios(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PortfolioWithAccounts>, String> {
    state
        .portfolio_service()
        .list_portfolios()
        .map_err(|e| format!("Failed to load portfolios: {}", e))
}

#[tauri::command]
pub async fn get_portfolio(
    portfolio_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioWithAccounts, String> {
    state
        .portfolio_service()
        .get_portfolio(&portfolio_id)
        .map_err(|e| format!("Failed to load portfolio: {}", e))
}

#[tauri::command]
pub async fn create_portfolio(
    portfolio: NewPortfolio,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioWithAccounts, String> {
    state
        .portfolio_service()
        .create_portfolio(portfolio)
        .await
        .map_err(|e| format!("Failed to create portfolio: {}", e))
}

#[tauri::command]
pub async fn update_portfolio_entry(
    portfolio: PortfolioUpdate,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioWithAccounts, String> {
    state
        .portfolio_service()
        .update_portfolio(portfolio)
        .await
        .map_err(|e| format!("Failed to update portfolio: {}", e))
}

#[tauri::command]
pub async fn delete_portfolio_entry(
    portfolio_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    state
        .portfolio_service()
        .delete_portfolio(&portfolio_id)
        .await
        .map_err(|e| format!("Failed to delete portfolio: {}", e))
}
